e2e/testcases/namespace_selectors_test.go

```diff
@@ -59,73 +100,264 @@ func TestNamespaceSelectorHierarchicalFormat(t *testing.T) {
 	nt.Must(nt.RootRepos[configsync.RootSyncName].Add("acme/cm-bookstore.yaml", bookstoreCM))
 	nt.Must(nt.RootRepos[configsync.RootSyncName].Add("acme/rq-bookstore.yaml", bookstoreRQ))
 	nt.Must(nt.RootRepos[configsync.RootSyncName].Add("acme/cm-shoestore.yaml", shoestoreCM))
 	nt.Must(nt.RootRepos[configsync.RootSyncName].CommitAndPush("Add Namespaces, NamespaceSelectors and Namespace-scoped resources"))
 
+	nt.Logger.Info("Only resources in shoestore are created because bookstore Namespace is not declared")
+	if err := nt.WatchForAllSyncs(); err != nil {
+		nt.T.Fatal(err)
+	}
+
+	validateSelectedAndUnselectedResources(nt,
+		append(selectedResourcesWithShoestoreNSSOnly),
+		append(unselectedResourcesWithShoestoreNSSOnly),
+	)
+	if err := nt.Validate(
+		configsync.RootSyncName,
+		configsync.ControllerNamespace,
+		&v1beta1.RootSync{},
+		testpredicates.MissingAnnotation(metadata.DynamicNSSelectorEnabledAnnotationKey),
+	); err != nil {
+		nt.T.Fatal(err)
+	}
+	if err := nt.Validate(
+		core.RootReconcilerName(configsync.RootSyncName),
+		configsync.ControllerNamespace,
+		&appsv1.Deployment{},
+		testpredicates.DeploymentMissingEnvVar(reconcilermanager.Reconciler, reconcilermanager.DynamicNSSelectorEnabled),
+	); err != nil {
+		nt.T.Fatal(err)
+	}
+
+	nt.Logger.Info("Update NamespaceSelector to use dynamic mode")
+	bookstoreNSS.Spec.Mode = v1.NSSelectorDynamicMode
+	nt.Must(nt.RootRepos[configsync.RootSyncName].Add("acme/namespace-selector-bookstore.yaml", bookstoreNSS))
+	nt.Must(nt.RootRepos[configsync.RootSyncName].CommitAndPush("Update NamespaceSelector to use dynamic mode"))
+	if err := nt.Watcher.WatchObject(kinds.RootSyncV1Beta1(),
+		configsync.RootSyncName,
+		configsync.ControllerNamespace,
+		[]testpredicates.Predicate{
+			testpredicates.HasAnnotation(
+				metadata.DynamicNSSelectorEnabledAnnotationKey, "true"),
+		}); err != nil {
+		nt.T.Fatal(err)
+	}
+	if err := nt.Watcher.WatchObject(kinds.Deployment(),
+		core.RootReconcilerName(configsync.RootSyncName),
+		configsync.ControllerNamespace,
+		[]testpredicates.Predicate{
+			testpredicates.DeploymentHasEnvVar(
+				reconcilermanager.Reconciler,
+				reconcilermanager.DynamicNSSelectorEnabled, "true"),
+		}); err != nil {
+		nt.T.Fatal(err)
+	}
+
+	if err := nt.WatchForAllSyncs(); err != nil {
+		nt.T.Fatal(err)
+	}
+	validateSelectedAndUnselectedResources(nt,
+		append(selectedResourcesWithShoestoreNSSOnly),
+		append(unselectedResourcesWithShoestoreNSSOnly),
+	)
+
+	nt.Logger.Info("Creating a Namespace with labels, matching resources should be selected")
+	bookstoreNamespace := fake.NamespaceObject(bookstoreNS, core.Label("app", bookstoreNS))
+	if err := nt.KubeClient.Create(bookstoreNamespace); err != nil {
+		nt.T.Fatal(err)
+	}
+	t.Cleanup(func() {
+		if err := nt.KubeClient.Delete(bookstoreNamespace); err != nil && !apierrors.IsNotFound(err) {
```

Comment:
I think you may need to disable the webhook here in case it's still running

Reply:
done. Also added a check for the webhook state to avoid re-stop.